### DISCUSSION TITLE
How to remote debugging on a vagrant machine?

### DISCUSSION THREAD

**Discussion by TornaxO7** [mfussenegger/nvim-dap#303] [Q&A] [answered]
Hello! I'm trying to remote debugging on my vagrant machine (which runs on my local machine).
This is in my Vagrantfile:

<details>

```
# ...
Vagrant.configure("2") do |config|
    config.vm.box = "generic/debian10"
    config.vm.network "forwarded_port", guest: 3128, host: 3128
    config.vm.network "forwarded_port", guest: 5678, host: 5678, protocol: "tcp", guest_ip: "127.0.0.1", host_ip: "127.0.0.1"
    # ...
end
```

</details>

I'd like to use the port `5678` for debugging.

This is my `~/.config/nvim/init.vim`:

<details>

```vim
call plug#begin('~/.vim/plugged')
" ...
Plug 'mfussenegger/nvim-dap'
Plug 'theHamsta/nvim-dap-virtual-text'
Plug 'rcarriga/nvim-dap-ui'
" ...
call plug#end()

" ...

source ~/.config/nvim/plugins/dap.vim
```

</details>

This is the `~/.config/nvim/plugins/dap.vim`:

<details>

```vim
augroup NvimDap
    autocmd!
    au FileType dap-repl lua require('dap.ext.autocompl').attach()
augroup END

let g:my_dap_repl_is_open = v:false

function OpenRepl()
    echo "called"
    " open repl if it isn't opened yet
    if !g:my_dap_repl_is_open
        lua require('dap').repl.open()
        let g:my_dap_repl_is_open = v:true
    endif

    lua require('dap').continue()
endfunction

nnoremap <F5> :call OpenRepl()<CR>
nnoremap <F6> :lua require('dap').step_over()<CR>
nnoremap <F7> :lua require('dap').step_into()<CR>
nnoremap <F8> :lua require('dap').step_out()<CR>
nnoremap <F9><F9> :lua require('dap').toggle_breakpoint()<CR>
nnoremap <F9><F10> :lua require('dap').set_breakpoint(vim.fn.input('Breakpoint condition: '))<CR>
nnoremap <F9><F7> :lua require('dap').set_breakpoint(nil, nil, vim.fn.input('Log point message: '))<CR>
"nnoremap <silent> <leader>dl :lua require'dap'.run_last()<CR>

lua require("dap_conf")
```

</details>

And this is in my `~/.config/nvim/lua/dap_conf.lua`:

<details>

```lua
local dap = require('dap')
local dap_ui = require('dapui')
dap.set_log_level('TRACE')
dap_ui.setup(
{
        icons = {
            expanded = "⯆",
            collapsed = "⯈",
            circular = "↺"
        },
        mappings = {
            expand = "<CR>",
            open = "o",
            remove = "d"
        },
        sidebar = {
            elements = {
                -- You can change the order of elements in the sidebar
                "scopes",
                "stacks",
                "watches"
            },
            width = 40,
            position = "left" -- Can be "left" or "right"
        },
        tray = {
            elements = {
                "repl"
            },
            height = 10,
            position = "bottom" -- Can be "bottom" or "top"
        },
        floating = {
            max_height = nil, -- These can be integers or a float between 0 and 1.
            max_width = nil   -- Floats will be treated as percentage of your screen.
        }
})

vim.g.dap_virtual_text = true

-- DAP CONFIG --
dap.adapters.python_local = {
    type = 'executable';
    command = 'python3';
    args = {'-m', 'debugpy.adapter'};
}

dap.adapters.python_remote = {
    type = 'server';
    host = '127.0.0.1';
    port = 5678;
}

dap.configurations.python = {
    {
        name = 'Remote file';
        type = 'python_remote';
        request = 'attach';
    },
    {
        name = 'Launch file';
        type = 'python_local';
        request = 'launch';
        program = '${file}';
    },
}
```

</details>

# Reproduction steps
1. ssh to the vagrant machine: `ssh vagrant@vagrant`
2. Go to your python file on the vagrant machine. For testing, use this file in `/tmp/test.py`:
```python
a = 10
b = 100

print(a + b)
```
3. Run `python3 -m debugpy --listen 5678 --wait-for-client /tmp/test.py`
4. On your local machine, do `nvim /tmp/test.py`
5. Run `:lua require('dap').continue()` (and use the `Remote` configuration of my config).

# Actual Behaviour
I'm getting the following output:

```
Debug adapter didn't respond. Either the adapter is slow (then wait and ignore this) or there is a problem with your adapter or `python_remote` configuration. Check the logs for errors (:help dap.set_log_level)
```

Since I set `dap.set_log_level('TRACE')`, this is in my `~/.cache/nvim/dap.log`:
```
[ DEBUG ] 2021-09-16T13:06:47Z+0200 ] /Users/tornax/.vim/plugged/nvim-dap/lua/dap/session.lua:750 ]	"request"	{
  arguments = {
    adapterID = "nvim-dap",
    clientId = "neovim",
    clientname = "neovim",
    columnsStartAt1 = true,
    linesStartAt1 = true,
    locale = "en_US.UTF-8",
    pathFormat = "path",
    supportsRunInTerminalRequest = true,
    supportsVariableType = true
  },
  command = "initialize",
  seq = 0,
  type = "request"
}
```

# Expected behaviour
A starting debugging session.

# Other information
If I run `python3 -m debugpy --listen 5678 --wait-for-client /tmp/test.py` locally on my machine, then I can debug just as usually but not on my vagrant machine ._.

**Comment by mfussenegger** [comment, marked as answer]
If you change the vagrant configuration to:

```
Vagrant.configure("2") do |config|
    config.vm.box = "generic/debian10"
    config.vm.network "forwarded_port", guest: 3128, host: 3128
    config.vm.network "forwarded_port", guest: 5678, host: 5678
    # ...
end
```

and start debugpy with ` python  -m debugpy --listen 0.0.0.0:5678 --wait-for-client /tmp/test.py` it should work.
Vagrant (or rather, virtualbox if using the default provider) sets up some 10.0.x.y interface for communication from host to guest system. If you start debugpy with `--listen 5678` it will bind to `127.0.0.1` in the guest and the host won't be able to communicate with it. Using `0.0.0.0:5678` ensures that it will bind the debugpy service on all interfaces.


Two other unrelated remarks:

- https://github.com/mfussenegger/nvim-dap-python would come with a single adapter definition and configurations for the local case and the remote case.
- There is already a `dap.repl.toggle()` function which you could use - no need to roll your own :)

**Reply by TornaxO7** [reply]
Yeah! I'm getting a different error message now! Thank you for the explanation with `127.0.0.1`! Now if I set a breakpoint in the first line of my `/tmp/test.py` then I'm getting the following error message:

```
Breakpoint in file that does not exist.
```

I could 'fix' this by adding the `breakpoint()` function to the same line as on my local machine. Is it possible to set a breakpoint according to my local debugging session or do I have to put always a `breakpoint()` function to the file which I'd like to debug?

**Reply by mfussenegger** [reply]
You'll probably have to configure the `pathMappings`, see https://github.com/microsoft/debugpy/wiki/Debug-configuration-settings#debugger-settings and https://github.com/microsoft/debugpy/wiki/Debugging-over-SSH

**Reply by TornaxO7** [reply]
Thank you sir! It's working now! :)

**Reply by TornaxO7** [reply]
I appreciate your help! 💯
